kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java

```diff
@@ -520,11 +283,34 @@ private TransactionCoordinator createAndBootTransactionCoordinator(String tenant
             TransactionCoordinator transactionCoordinator =
                     initTransactionCoordinator(tenant, brokerService.getPulsar().getAdminClient(), clusterData);
             // Listening transaction topic load/unload
-            brokerService.pulsar()
-                    .getNamespaceService()
-                    .addNamespaceBundleOwnershipListener(
-                            new TransactionStateRecover(brokerService, tenant, kafkaConfig, transactionCoordinator));
+            final NamespaceName kafkaMetaNs = NamespaceName.get(tenant, kafkaConfig.getKafkaMetadataNamespace());
+            final NamespaceName kafkaTopicNs = NamespaceName.get(tenant, kafkaConfig.getKafkaNamespace());
+            final String metadataNamespace = kafkaConfig.getKafkaMetadataNamespace();
+            bundleListener.addTopicOwnershipListener(new TopicOwnershipListener() {
+                @Override
+                public void whenLoad(TopicName topicName) {
+                    if (KopTopic.isTransactionMetadataTopicName(topicName.toString(), metadataNamespace)) {
+                        transactionCoordinator.handleTxnImmigration(topicName.getPartitionIndex());
+                    }
+                }
 
+                @Override
+                public void whenUnload(TopicName topicName) {
+                    if (KopTopic.isTransactionMetadataTopicName(topicName.toString(), metadataNamespace)) {
+                        transactionCoordinator.handleTxnEmigration(topicName.getPartitionIndex());
+                    }
+                }
+
+                @Override
+                public String name() {
+                    return "TransactionStateRecover";
+                }
+
+                @Override
+                public boolean test(NamespaceName namespaceName) {
+                    return namespaceName.equals(kafkaMetaNs) || namespaceName.equals(kafkaTopicNs);
+                }
+            });
```

Comment:
In PR #964, we expect `metadata topics` to appear within the `metadata namespace`. For the `test` method, we do not need to judge the `kafkaNamespace`.

Reply:
Sounds reasonable. It looks like this redundant check has been here from the beginning when `OffsetAndTopicListener` and `TransactionStateRecover` were added.